### PR TITLE
Add implementations of Device for hal and ash

### PR DESCRIPTION
**Pull request by magnonellie**
There are probably a lot of mistakes and missteps everywhere here, but it can probably be worked on at least :stuck_out_tongue:

### CODE REVIEW

File: resource/src/impls/hal.rs [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+    unsafe fn destroy_buffer(&self, buffer: Self::Buffer) {
+        self.0.borrow().destroy_buffer(buffer);
+    }
+
+    fn create_image(
+        &self,
+        info: image::CreateInfo,
+    ) -> Result<Self::UnboundImage, error::ImageCreationError> {
+        let kind = match info.kind {
+            image::Kind::D1 => hal::image::Kind::D1(info.extent.width, 0),
```
Comment (msiglreith) [inline resource/src/impls/hal.rs]: should be 1 layer not 0

File: resource/src/impls/hal.rs [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+            image::Kind::D2 => hal::image::Kind::D2(
+                info.extent.width,
+                info.extent.height,
+                0,
```
Comment (msiglreith) [inline resource/src/impls/hal.rs]: should be 1 layer not 0

File: resource/src/impls/hal.rs [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+                hal::image::Kind::D3(info.extent.width, info.extent.height, info.extent.depth)
+            }
+        };
+        let format = to_hal_format(info.format);
+        let tiling = match info.tiling {
+            image::ImageTiling::Optimal => hal::image::Tiling::Optimal,
+            image::ImageTiling::Linear => hal::image::Tiling::Linear,
+        };
+        let usage = hal::image::Usage::from_bits_truncate(info.usage.bits());
+        let view_caps = hal::image::ViewCapabilities::all(); // I really have no idea what this is. Metal maybe?
```
Comment (msiglreith) [inline resource/src/impls/hal.rs]: these are flags exposed by vulkan indicating which image views parameters may be set. It's a subset of https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkImageCreateFlagBits.html

Comment (magnonellie) [reply resource/src/impls/hal.rs]: Are they required or can you leave them empty? If they are required, how would I go about finding the correct flags here?

Comment (msiglreith) [reply resource/src/impls/hal.rs]: * `mutable`: Want to create a view with a different format than specified on image creation, should be avoided if possible.
* `cube`: if you want to create an cube view
* `2d_array`: View a 3D Image as 2D image array, or a slice of it, or one layer as 2D image view

File: resource/src/impls/hal.rs (hunk 1 above)
Comment (magnonellie) [reply resource/src/impls/hal.rs]: These are referring to array layers, aren't they?

File: command/src/device.rs [hunk 4]
```diff
@@ -14,30 +14,28 @@ pub trait Device: resource::Device {
     type Fence: Debug + 'static;
 
     /// Finished command buffer that can be submitted to the queues of this device.
-    type Submit: Debug + 'static;
```
Comment (omni-viral) [inline command/src/device.rs]: Why did you remove `Debug` bound?

File: command/src/impls/ash.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+{
+    type Semaphore = vk::Semaphore;
+    type Fence = vk::Fence;
+    type Submit = ();
+    type CommandPool = vk::CommandPool;
+    type CommandBuffer = vk::CommandBuffer;
+    type CommandQueue = vk::Queue;
+}
+
+impl CommandBuffer for vk::CommandBuffer {
```
Comment (omni-viral) [inline command/src/impls/ash.rs]: Function pointers are required since you going to call recording methods with it.
`(vk::DeviceFnV1_0, vk::CommandBuffer)` should do.

File: command/src/impls/ash.rs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    type Submit = ();
```
Comment (omni-viral) [inline command/src/impls/ash.rs]: `type Submit = vk::CommandBuffer;`

File: command/src/impls/hal.rs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+use device::{CommandBuffer, CommandQueue, Device};
+
+impl<D, B> Device for (D, PhantomData<B>)
+where
+    B: hal::Backend,
+    D: Borrow<B::Device>,
+{
+    type Semaphore = B::Semaphore;
+    type Fence = B::Fence;
+    type Submit = ();
```
Comment (omni-viral) [inline command/src/impls/hal.rs]: `type Submit = B::CommandBuffer;`

File: command/src/impls/hal.rs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+impl<C, B> CommandBuffer for (C, PhantomData<B>)
+    where
+        B: hal::Backend,
+        C: Borrow<B::CommandBuffer>,
+{
+    type Submit = ();
+
+    unsafe fn submit(&self) -> Self::Submit {
+        unimplemented!()
```
Comment (omni-viral) [inline command/src/impls/hal.rs]: `self.clone()`

File: resource/src/impls/ash.rs [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+            Astc10x6SrgbBlock => Format::ASTC_10x6_SRGB_BLOCK,
+            Astc10x8UnormBlock => Format::ASTC_10x8_UNORM_BLOCK,
+            Astc10x8SrgbBlock => Format::ASTC_10x8_SRGB_BLOCK,
+            Astc10x10UnormBlock => Format::ASTC_10x10_UNORM_BLOCK,
+            Astc10x10SrgbBlock => Format::ASTC_10x10_SRGB_BLOCK,
+            Astc12x10UnormBlock => Format::ASTC_12x10_UNORM_BLOCK,
+            Astc12x10SrgbBlock => Format::ASTC_12x10_SRGB_BLOCK,
+            Astc12x12UnormBlock => Format::ASTC_12x12_UNORM_BLOCK,
+            Astc12x12SrgbBlock => Format::ASTC_12x12_SRGB_BLOCK,
+        }
```
Comment (omni-viral) [inline resource/src/impls/ash.rs]: I hope your fingers don't hurt after typing this 😅

File: resource/src/impls/ash.rs [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+            sharing_mode: match info.sharing {
+                SharingMode::Exclusive => vk::SharingMode::Exclusive,
+            },
+            queue_family_index_count: 0,
+            p_queue_family_indices: ptr::null(),
+        };
+
+        unsafe {
+            <Self as DeviceV1_0>::create_buffer(self, &info, None)
+        }.map_err(|_| memory::OutOfMemoryError::OutOfDeviceMemory)
```
Comment (omni-viral) [inline resource/src/impls/ash.rs]: You should match concrete error variants and panic on unexpected error variants.

File: resource/src/impls/ash.rs [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+        }
+    }
+
+    unsafe fn bind_buffer(
+        &self,
+        buffer: Self::UnboundBuffer,
+        memory: &Self::Memory,
+        offset: u64,
+    ) -> Result<Self::Buffer, error::BindError> {
+        <Self as DeviceV1_0>::bind_buffer_memory(self, buffer, *memory, offset)?;
```
Comment (omni-viral) [inline resource/src/impls/ash.rs]: Can just write `DeviceV1_0::bind_buffer_memory(self...`

File: resource/src/impls/hal.rs (hunk 1 above)
Comment (omni-viral) [reply resource/src/impls/hal.rs]: Correct

File: command/src/device.rs (hunk 4 above)
Comment (magnonellie) [reply command/src/device.rs]: HAL's command buffers aren't `Debug` unfortunately.

Comment (omni-viral) [reply command/src/device.rs]: Indeed. Well, they should be.

File: resource/src/error.rs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+/// Image creation error.
+#[derive(Clone, Copy, Debug, Fail)]
+pub enum ImageCreationError {
+    /// An unsupported format was attempted to be used.
+    #[fail(display = "Unsupported format")]
+    UnsupportedFormat(image::Format),
+
+    /// The kind cannot support a particular operation.
+    #[fail(display = "Unsupported kind")]
+    Kind,
```
[the remark this answers is not in the document]
Comment (magnonellie) [reply resource/src/error.rs]: I honestly have no idea what most of these errors even are. I pretty much just copied them from `hal` :sweat: 

What exactly do you want me to change with this?

File: resource/src/impls/hal.rs [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+            image::Kind::D3 => {
+                hal::image::Kind::D3(info.extent.width, info.extent.height, info.extent.depth)
+            }
+        };
+        let format = to_hal_format(info.format);
+        let tiling = match info.tiling {
+            image::ImageTiling::Optimal => hal::image::Tiling::Optimal,
+            image::ImageTiling::Linear => hal::image::Tiling::Linear,
+        };
+        let usage = hal::image::Usage::from_bits_truncate(info.usage.bits());
```
[the remark this answers is not in the document]
Comment (magnonellie) [reply resource/src/impls/hal.rs]: Would an unwrap be okay or are we staying away from those?

File: resource/src/resources.rs [hunk 14]
```diff
@@ -25,12 +26,14 @@ impl<T: 'static, B: 'static, I: 'static> Resources<T, B, I> {
         info: buffer::CreateInfo,
         align: u64,
         memory_usage: M,
-    ) -> Result<buffer::Buffer<T, B>, MemoryError>
+    ) -> Result<buffer::Buffer<T, B>, ResourceError>
```
[the remark this answers is not in the document]
Comment (magnonellie) [reply resource/src/resources.rs]: It needs to be `ResourceError` though since it can emit either a `BindError` or `MemoryError`.

File: resource/src/error.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (omni-viral) [reply resource/src/error.rs]: From `gfx-hal` sources I see that this error means that "Multi-sampled array textures or cubes are not supported" by metal backend. Just  change the comment.

File: resource/src/impls/hal.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (omni-viral) [reply resource/src/impls/hal.rs]: I follow this simple algorithm to decide:
* Handle if error is expected
* `expect` if error is expected but handling is meaningless.
* `unwrap` if I'm sure that it won't panic.
* And
```rust
unsafe {
  match value {
    Some(value) => value,
    None => ::std::hint::unreachable_unchecked(),
  }
}
```
when I'm 200% sure

File: resource/src/resources.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (omni-viral) [reply resource/src/resources.rs]: No. It shouldn't be able to emit `BindError`.
Binding function should always be successful.

**Comment by omni-viral** [comment]
Great!
Let's merge it.
